raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java

```diff
@@ -0,0 +1,271 @@
+
+        // check if leader response were to contain bootstrap snapshot id, follower would not send fetch snapshot request
+        context.deliverResponse(
+            fetchRequest.correlationId(),
+            fetchRequest.destination(),
+            context.snapshotFetchResponse(epoch, leader, BOOTSTRAP_SNAPSHOT_ID, 0)
+        );
+        context.pollUntilRequest();
+        fetchRequest = context.assertSentFetchRequest();
+        context.assertFetchRequestData(fetchRequest, epoch, 0L, 0);
+    }
+
+    // this test is maybe out of scope - we are checking that follower is able to apply votersRecord essentially
+    @ParameterizedTest
+    @ValueSource(booleans = { true, false })
+    public void testFollowerReadsKRaftBootstrapRecords(boolean withKip853Rpc) throws Exception {
+        int localId = 0;
+        int leader = 1;
+        int follower2 = 2;
+        Uuid localDirectoryId = Uuid.randomUuid();
+        Uuid leaderDirectoryId = Uuid.randomUuid();
+        Uuid followerDirectoryId2 = Uuid.randomUuid();
+        Set<Integer> voterIds = new HashSet<>(Arrays.asList(localId, leader));
+        Set<ReplicaKey> voters = new HashSet<>(Arrays.asList(
+            ReplicaKey.of(localId, localDirectoryId),
+            ReplicaKey.of(leader, leaderDirectoryId)));
+        VoterSet voterSet = VoterSetTest.voterSet(voters.stream());
+        short kRaftVersion = 1;
+        int epoch = 5;
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(OptionalInt.of(localId), voterIds, localDirectoryId, kRaftVersion)
+            .withKip853Rpc(withKip853Rpc)
+            .withElectedLeader(epoch, leader)
+            .withBootstrapSnapshot(Optional.of(voterSet), kRaftVersion)
+            .appendToLog(epoch, Arrays.asList("a", "b", "c"))
+            .build();
+
+        // check that follower will send fetch request to leader
+        context.pollUntilRequest();
+        RaftRequest.Outbound fetchRequest = context.assertSentFetchRequest();
+        context.assertFetchRequestData(fetchRequest, epoch, 3L, 5);
+
+        // leader sends batch with bootstrap records
+        Set<ReplicaKey> leadersVoters = new HashSet<>(voters);
+        leadersVoters.add(ReplicaKey.of(follower2, followerDirectoryId2));
+        VoterSet leadersVoterSet = VoterSetTest.voterSet(leadersVoters.stream());
+        List<String> leaderRecords = Arrays.asList(leadersVoterSet.toVotersRecord(ControlRecordUtils.KRAFT_VOTERS_CURRENT_VERSION).toString());
+        MemoryRecords batch = context.buildControlBatch(3L, epoch, leaderRecords);
+        context.deliverResponse(
+            fetchRequest.correlationId(),
+            fetchRequest.destination(),
+            context.fetchResponse(epoch, leader, batch, 3, Errors.NONE));
+
+        // this is broken because RaftClientTestContext.buildBatch does not send records in viable format
+        // follower applies the bootstrap records, registering follower2 as a new voter
```

Comment:
Why is this broken? What do you mean by viable format?

Reply:
When the follower attempts to load the log with these records it throws an NPE because the record.key() is null

Reply:
thanks for discussing this offline, I'll try appending control records differently